core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -1517,6 +1517,8 @@ private static Object clean(Object o, RelDataType type) {
       if (o instanceof BigDecimal) {
         return o;
       }
+      assert !(o instanceof Float || o instanceof Double)
+          : o.getClass().getCanonicalName() + " is not compatible with " + type.getSqlTypeName();
       return new BigDecimal(((Number) o).longValue());
```

Comment:
What do you mean by "s not compatible with" ?

Reply:
Will it better to include a reminder that use `makeExactLiteral` to create Decimal literal by a float or double input?